### PR TITLE
Implement _most_ of the EA interface for DTA/TDA

### LINKED ISSUE #23586: PERF: Implement PeriodArray._unique
Avoid an astype(object).

```diff
diff --git a/pandas/core/arrays/period.py b/pandas/core/arrays/period.py
index 5a75f2706..12191ad89 100644
--- a/pandas/core/arrays/period.py
+++ b/pandas/core/arrays/period.py
@@ -216,6 +216,10 @@ class PeriodArray(dtl.DatetimeLikeArrayMixin, ExtensionArray):
         ordinals = libperiod.extract_ordinals(periods, freq)
         return cls(ordinals, freq=freq)
 
+    def unique(self):
+        from pandas.core.algorithms import unique
+        return type(self)(unique(self.asi8), self.freq)
+
     def _values_for_factorize(self):
         return self.asi8, iNaT
```

should work.

### PR DESCRIPTION
**Pull request by jbrockmendel**
@TomAugspurger as promised, this implements a handful of tests, and a few more can be ported from #23415, but the EA-specific tests haven't been started.

The big missing piece is DatetimeArray._from_sequence, which I'm getting started on now in a new branch.

Closes #23586

### CODE REVIEW

File: pandas/core/indexes/datetimes.py [hunk 1]
```diff
@@ -551,16 +551,12 @@ def snap(self, freq='S'):
...
-        if self.tz is not None:
-            naive = type(self)(self._ndarray_values, copy=False)
-        else:
-            naive = self
-        result = super(DatetimeIndex, naive).unique(level=level)
-        return self._shallow_copy(result.values)
+        if level is not None:
+            self._validate_index_level(level)
+
+        # call DatetimeArray method
```
Comment (TomAugspurger) [inline pandas/core/indexes/datetimes.py]: Could you start adding comments like `TODO(DatetimeArray)` here? Once we switch to composition, we should be able to remove the PeriodIndex implementation and use this.

File: pandas/core/arrays/datetimelike.py [hunk 2]
```diff
@@ -196,6 +196,71 @@ def astype(self, dtype, copy=True):
...
         return super(DatetimeLikeArrayMixin, self).astype(dtype, copy)
 
+    # ------------------------------------------------------------------
+    # ExtensionArray Interface
+    # TODO:
+    #   * _from_sequence
+    #   * argsort / _values_for_argsort
+    #   * _reduce
+
+    def unique(self):
```
Comment (TomAugspurger) [inline pandas/core/arrays/datetimelike.py]: I suppose this closes  https://github.com/pandas-dev/pandas/issues/23586 then? Can you confirm and add a "closes #23586" to the original issue?

File: pandas/core/arrays/datetimelike.py [hunk 3]
```diff
@@ -196,6 +196,71 @@ def astype(self, dtype, copy=True):
...
+        # for TimedeltaArray and PeriodArray; DatetimeArray overrides
+        freqs = {x.freq for x in to_concat}
+        assert len(freqs) == 1
+        freq = list(freqs)[0]
+        values = np.concatenate([x.asi8 for x in to_concat])
+        return cls._simple_new(values, freq=freq)
+
+    def copy(self, deep=False):
+        values = self.asi8
+        if deep:
```
Comment (TomAugspurger) [inline pandas/core/arrays/datetimelike.py]: @jorisvandenbossche does this look like the correct behavior to you? I was thinking that `.copy()` should break a chain such that in

```python
a = DatetimeArray(...)
b = a.copy()
a[0] = np.nan
```

`b[0]` should not be updated along with `a[0]`. IIUC, with the current implementation `a` and `b` would have the same underlying ndarray (with `deep=False`).

In my head, the `deep` parameter is really only useful when the "scalars" are actually containers (nested data). In this case, `deep=True` would mean recursively copy the values in the "scalar" containers.

File: pandas/core/arrays/datetimelike.py [hunk 4]
```diff
@@ -196,6 +196,71 @@ def astype(self, dtype, copy=True):
...
+            values = values.copy()
+        return type(self)(values, dtype=self.dtype, freq=self.freq)
+
+    def _values_for_factorize(self):
+        return self.asi8, iNaT
+
+    @classmethod
+    def _from_factorized(cls, values, original):
+        if is_datetime64tz_dtype(original):
```
Comment (TomAugspurger) [inline pandas/core/arrays/datetimelike.py]: Can this `if` be replaced with `cls(values, dtype=original.dtype, freq=original.freq)`? `original.dtype` should have the tzinfo in that case.

File: pandas/core/arrays/datetimelike.py [hunk 5]
```diff
@@ -196,6 +196,71 @@ def astype(self, dtype, copy=True):
...
+        new_values = take(self.asi8,
+                          indices,
+                          allow_fill=allow_fill,
+                          fill_value=fill_value)
+
+        return type(self)(new_values, dtype=self.dtype)
+
+    @classmethod
+    def _concat_same_type(cls, to_concat):
+        # for TimedeltaArray and PeriodArray; DatetimeArray overrides
```
Comment (TomAugspurger) [inline pandas/core/arrays/datetimelike.py]: The problem being that DatetimeArray needs to pass through `tz` info?

For PeriodArray at least (haven't checked TimedeltaArray) you should able to implement `_concat_same_type` just in terms of `.dtype`. It's hashable and can be passed to `PeriodArray.__init__`.

File: pandas/core/arrays/timedeltas.py [hunk 6]
```diff
@@ -139,7 +140,7 @@ def _simple_new(cls, values, freq=None, dtype=_TD_DTYPE):
         result._freq = freq
         return result
 
-    def __new__(cls, values, freq=None):
+    def __new__(cls, values, freq=None, dtype=_TD_DTYPE):
```
Comment (TomAugspurger) [inline pandas/core/arrays/timedeltas.py]: (why) was this change necessary?

File: pandas/core/arrays/datetimelike.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: The `if deep:` branch here was based on jeff's comment [here](https://github.com/pandas-dev/pandas/pull/23415#discussion_r232497536).  I haven't thought about this enough to have a strong opinion about it other than that we shouldn't silently ignore the `deep` kwarg.

File: pandas/core/arrays/datetimelike.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: Good call

File: pandas/core/arrays/datetimelike.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: Will do.

File: pandas/core/arrays/datetimelike.py (hunk 3 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/datetimelike.py]: Or simply remove `deep` from the interface? (but that can maybe discussed elsewhere, as it is already part of the interface currenlty)

In any case, I think copy, even with deep=False (the default), should clearly copy the values so that modifying the copy does not change the original one like Tom showed in the example.

File: pandas/core/arrays/timedeltas.py (hunk 6 above)
Comment (jbrockmendel) [reply pandas/core/arrays/timedeltas.py]: Because in the copy, from_factorized, concat_same_type methods we are passing dtype to this

File: pandas/core/arrays/datetimelike.py (hunk 5 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/datetimelike.py]: As opposed to PeriodArray, `freq` is not part of the dtype for DatetimeArray/TimedeltaArray, so I am not sure this check for freq should be done here.

File: pandas/core/arrays/datetimelike.py (hunk 4 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/datetimelike.py]: I think this `if` can simply be removed? The values to be factorized are explicitly integers, so `_from_factorized` should always be passed integers.

File: pandas/core/arrays/timedeltas.py (hunk 6 above)
Comment (TomAugspurger) [reply pandas/core/arrays/timedeltas.py]: Whoops, I completely missed that the `dtype` argument wasn't present there before. I thought you were changing it from None to `_TD_DTYPE`.

File: pandas/tests/arrays/test_datetimelike.py [hunk 7]
```diff
@@ -56,7 +56,68 @@ def timedelta_index(request):
     return pd.TimedeltaIndex(['1 Day', '3 Hours', 'NaT'])
 
 
-class TestDatetimeArray(object):
+class SharedTests(object):
+    index_cls = None
+
+    def test_take(self):
```
Comment (jorisvandenbossche) [inline pandas/tests/arrays/test_datetimelike.py]: We should not need to add such basic tests I think, as those are covered by the base Extension tests (we should of course test datetime specific aspects).

Is there anything in this test not tested by the base tests?

File: pandas/tests/arrays/test_datetimelike.py [hunk 8]
```diff
@@ -56,7 +56,68 @@ def timedelta_index(request):
...
+            arr.take([0, 1], allow_fill=True, fill_value=2)
+
+        with pytest.raises(ValueError):
+            arr.take([0, 1], allow_fill=True, fill_value=2.0)
+
+        with pytest.raises(ValueError):
+            arr.take([0, 1], allow_fill=True,
+                     fill_value=pd.Timestamp.now().time)
+
+    def test_concat_same_type(self):
```
Comment (jorisvandenbossche) [inline pandas/tests/arrays/test_datetimelike.py]: As I think also commented on one of the previous PRs that started doing this, I don't think we should test `_concat_same_type` here directly. It is already tested by the base extension tests and by all the tests that actually use it under the hood.

File: pandas/core/arrays/datetimelike.py [hunk 9]
```diff
@@ -196,6 +196,74 @@ def astype(self, dtype, copy=True):
...
+        if deep:
+            values = values.copy()
+        return type(self)(values, dtype=self.dtype, freq=self.freq)
+
+    def _values_for_factorize(self):
+        return self.asi8, iNaT
+
+    @classmethod
+    def _from_factorized(cls, values, original):
+        return cls(values, dtype=original.dtype, freq=original.freq)
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/datetimelike.py]: Does it make sense to pass `original.freq` here? 
It seems a bit strange as this is creating a new array which does not necessarily have the same order as the original one.

Although in practice, if you *have* a freq, that means you have a regular and unique array to start with, so the factorization is kind of a no-op and the result will still have the same freq? (but might be missing corner cases)

Comment (jorisvandenbossche) [reply pandas/core/arrays/datetimelike.py]: Thinking of a possible corner case, which is currently actually broken: a sorted factorize of a DatetimeIndex with a negative freq: 

```
In [57]: idx = pd.date_range("2012-01-01", periods=3)

In [58]: idx
Out[58]: DatetimeIndex(['2012-01-01', '2012-01-02', '2012-01-03'], dtype='datetime64[ns]', freq='D')

In [59]: pd.factorize(idx)
Out[59]: 
(array([0, 1, 2]),
 DatetimeIndex(['2012-01-01', '2012-01-02', '2012-01-03'], dtype='datetime64[ns]', freq='D'))

In [60]: pd.factorize(idx[::-1])
Out[60]: 
(array([0, 1, 2]),
 DatetimeIndex(['2012-01-03', '2012-01-02', '2012-01-01'], dtype='datetime64[ns]', freq='-1D'))

In [61]: pd.factorize(idx[::-1], sort=True)
Out[61]: 
(array([2, 1, 0]),
 DatetimeIndex(['2012-01-01', '2012-01-02', '2012-01-03'], dtype='datetime64[ns]', freq='-1D'))
```

File: pandas/tests/arrays/test_datetimelike.py (hunk 7 above)
Comment (jbrockmendel) [reply pandas/tests/arrays/test_datetimelike.py]: The class-specific tests have tests for invalid fill_values

File: pandas/tests/arrays/test_datetimelike.py (hunk 8 above)
Comment (jbrockmendel) [reply pandas/tests/arrays/test_datetimelike.py]: Yah, most of these tests were salvaged from one of those older PRs.  I don't see much downside to having the tests, but am pretty happy to pawn this decision/PR off on Tom

**Comment by jbrockmendel** [comment]
@TomAugspurger are you still good taking this PR over?  (BTW Travis failures look like unrelated timeouts)

**Comment by TomAugspurger** [comment]
Sure will push changes here.

File: pandas/core/arrays/datetimelike.py (hunk 3 above)
Comment (TomAugspurger) [reply pandas/core/arrays/datetimelike.py]: We have https://github.com/pandas-dev/pandas/issues/22314 for the meaning of `deep`.

File: pandas/core/arrays/datetimelike.py (hunk 5 above)
Comment (TomAugspurger) [reply pandas/core/arrays/datetimelike.py]: IIRC, the PeriodArray constructor allows duplicate `freq` and `dtype`, as long as the match, so

```python
PeriodArray(data, freq='H', dtype=PeriodDtype("H")
```

should be fine.

Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: Yah, we added dtype to the PeriodArray constructor specifically so that type(self)(values, freq=self.freq, dtype=self.dtype) would be valid for all three TDA/DTA/PA classes

File: pandas/tests/arrays/test_datetimelike.py (hunk 8 above)
Comment (TomAugspurger) [reply pandas/tests/arrays/test_datetimelike.py]: We still want the fail causes (different dtypes) here.

At this point, we should be able to simplify `core/dtypes/concat.py::concat_datetimetz` and `DatetimeIndex._concat_same_dtype` right? I'll take a look.

File: pandas/core/arrays/datetimelike.py (hunk 5 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/datetimelike.py]: Sorry, that's not what I was meaning. The difference is the meaning of `freq` for PeriodArray vs DatetimeArray. For PeriodArray it is part of the dtype and it is defining how the stored integers are interpreted (and thus need to match to just concatenate those integers), but for DatetimeArray it is simply an informative attribute telling you about the regularity of the Array, but not essential to describe it. So I would assume that `_concat_same_type` needs to handle different arrays with different freqs for DatetimeArray.

**Comment by TomAugspurger** [comment]
Ah, I see what you mean now. Yes you're right.

And I suppose we can't even assume that concatenating multiple
DatetimeArrays with the same
freq will end up with a DatetimeIndex with the same freq.

**Comment by jorisvandenbossche** [comment]
> And I suppose we can't even assume that concatenating multiple DatetimeArrays with the same freq will end up with a DatetimeIndex with the same freq.

Indeed, as that would only be the case if they are nicely consecutive ranges.

**Comment by TomAugspurger** [comment]
Pushed some changes.

1. DatetimelikeArray.copy now copies the underlying data
2. changed the concatenation of DatetimeIndex, hopefully for the better.

File: pandas/core/arrays/timedeltas.py (hunk 6 above)
Comment (jreback) [reply pandas/core/arrays/timedeltas.py]: shouldn't this not still be passed as None and an if check inside? e.g.

```
if dtype is None:
     dtype = _TD_DTYPE
```

as seems passing None should error now

File: pandas/core/indexes/datetimelike.py [hunk 10]
```diff
@@ -685,17 +689,21 @@ def _concat_same_dtype(self, to_concat, name):
...
 
         if not is_period_dtype(self):
             # reset freq
             attribs['freq'] = None
-
-        if getattr(self, 'tz', None) is not None:
-            return _concat._concat_datetimetz(to_concat, name)
+            # TODO(DatetimeArray)
+            # - remove the .asi8 here
+            # - remove the _maybe_box_as_values
```
Comment (jreback) [inline pandas/core/indexes/datetimelike.py]: looks like you can remove part of the comment

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
just a really minor comment. ok to merge.

**Review by jreback — APPROVED** [review]

**Comment by TomAugspurger** [comment]
Going to leave the stale comment for now if that's OK. That whole if / else should be going away soon.

**Comment by jreback** [comment]
ok! thanks @jbrockmendel and @TomAugspurger
